### PR TITLE
fix(layout): preserve contextual digit runs

### LINKED ISSUE #200: Digit-only text runs are silently dropped from markdown output
### Summary

When a number is emitted as its own text-showing operation (its own `Tj`), it is silently removed from the markdown produced by `processPdf` / `pdf2md`. The surrounding words are kept and the sentence stays grammatical, so the loss leaves no trace in the output.

`extractTextWithPositions` returns the number correctly, so the content survives extraction — it disappears during markdown conversion.

### Why this matters

The failure is invisible. A sentence like

> a total of **730** seats and **17** places, up to a cut of **5** points

comes out as

> a total of seats and places, up to a cut of points

which still reads as valid prose. A consumer has no signal that anything was lost — no placeholder, no warning, no `needsOcr`, and `confidence` stays at `1.0`.

This is not a rare layout. Word and LibreOffice emit **one `Tj` per word** for justified text, so ordinary office PDFs hit it constantly. We found it on a public institutional document where the dropped tokens were the quantities the document exists to communicate, and on a school exam where `log 2 = a and log 5 = c` became `log = a and log = c`.

### Reproduction

Attached: `repro.pdf` (1 KB, 3 lines, generated by the script below — no third-party content).

```
node -e "const{readFileSync}=require('fs');console.log(require('@firecrawl/pdf-inspector').processPdf(readFileSync('repro.pdf')).markdown)"
```

**Actual:**

```
A) The total of seats was approved.
B) let log = a
C) Control: The total of 730 seats was approved. let log 2 = a
```

**Expected:**

```
A) The total of 730 seats was approved.
B) let log 2 = a
C) Control: The total of 730 seats was approved. let log 2 = a
```

Lines A and B place every word as a separate `Tj`. Line C is the same text as a single `Tj` and is unaffected.

`extractTextWithPositions` on the same file returns all three lines complete, including `730` and `2`:

```
["A)","The","total","of","730","seats","was","approved.",
 "B)","let","log","2","=","a",
 "C) Control: The total of 730 seats was approved. let log 2 = a"]
```

### Scope observed

- Triggers when the run contains **only digits**.
- Does **not** trigger when the digits share a run with other characters (`(730)`, `730,`, or the whole sentence in one run).
- Independent of digit count (`5`, `17`, `730`, `2026` all reproduce) and of the horizontal gap around the run.
- Does **not** reproduce when the layout analyzer routes the page through table detection — in table output the numbers are preserved. So it appears to be in the plain-text markdown path.

### Generator for the attached PDF

```python
#!/usr/bin/env python3
LINES = [
    (780, [(72,'A)'),(96,'The'),(126,'total'),(166,'of'),(186,'730'),(220,'seats'),(262,'was'),(296,'approved.')]),
    (755, [(72,'B)'),(96,'let'),(120,'log'),(150,'2'),(164,'='),(180,'a')]),
    (720, [(72,'C) Control: The total of 730 seats was approved. let log 2 = a')]),
]

def esc(s):
    # PDF literal strings must escape \, ( and ) — the runs above contain ')'
    return s.replace('\\', '\\\\').replace('(', '\\(').replace(')', '\\)')

ops = ["BT", "/F1 12 Tf"]
for y, runs in LINES:
    for x, txt in runs:
        ops.append(f"1 0 0 1 {x} {y} Tm ({esc(txt)}) Tj")
ops.append("ET")
stream = "\n".join(ops).encode("latin-1")

objs = [
    b"<< /Type /Catalog /Pages 2 0 R >>",
    b"<< /Type /Pages /Kids [3 0 R] /Count 1 >>",
    b"<< /Type /Page /Parent 2 0 R /MediaBox [0 0 595 842] "
    b"/Resources << /Font << /F1 5 0 R >> >> /Contents 4 0 R >>",
    b"<< /Length " + str(len(stream)).encode() + b" >>\nstream\n" + stream + b"\nendstream",
    b"<< /Type /Font /Subtype /Type1 /BaseFont /Helvetica >>",
]

out = bytearray(b"%PDF-1.4\n")
offsets = []
for i, body in enumerate(objs, start=1):
    offsets.append(len(out))
    out += f"{i} 0 obj\n".encode() + body + b"\nendobj\n"
xref = len(out)
out += f"xref\n0 {len(objs)+1}\n".encode() + b"0000000000 65535 f \n"
for off in offsets:
    out += f"{off:010d} 00000 n \n".encode()
out += f"trailer\n<< /Size {len(objs)+1} /Root 1 0 R >>\nstartxref\n{xref}\n".encode() + b"%%EOF\n"
open("repro.pdf", "wb").write(bytes(out))
```

### Environment

- `@firecrawl/pdf-inspector` 1.11.2 (npm)
- Reproduced on Linux x86_64 (`linux-x64-gnu` binary) and on Windows x64 (`win32-x64-msvc` binary), Node.js 24.18.0

[repro.pdf](https://github.com/user-attachments/files/30627426/repro.pdf)

### PR DESCRIPTION
**Pull request by abimaelmartell**
## Summary

Preserve digit-only text operations when they are geometrically attached to surrounding line content, while continuing to remove standalone and recurring page folios before layout analysis.

Adds regression coverage for separate numeric text operations near page edges. Fixes #200.

---
<a href="https://app.blacksmith.sh/firecrawl/codesmith/pdf-inspector/pr/201"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-dark-v2.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-light-v2.svg">View with [code]smith</picture></a> <a href="https://backend.blacksmith.sh/track/enable-autofix?expires=1788246261&installation_model_id=11126&pr_number=201&repository=firecrawl%2Fpdf-inspector&return_to=https%3A%2F%2Fgithub.com%2Ffirecrawl%2Fpdf-inspector%2Fpull%2F201&signature=2c530fba4f9e75f5d2c7aa7ca16a8fb314e7349bbf037916eba0a942813a46ee"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-light.svg">Autofix with [code]smith</picture></a>
<sup>Need help on this PR? Tag <code>@codesmith-bot</code> with what you need. Autofix is disabled.</sup>

---
## Summary by cubic
Preserves contextual digit-only text while removing running folios earlier using document‑wide, spread, and adjacent‑page signals. Prevents folios from skewing layout and fixes widget page mapping in forms. Fixes #200.

- **Bug Fixes**
  - Detects and removes running folios with document‑wide evidence; supports spreads and adjacent-page alternation; constrained to page‑edge zones with content‑width floors and tighter baselines.
  - Preserves contextual digits: requires nearby non‑numeric within ≤1.5em; numeric candidates don’t contextualize each other; keeps dense numeric structures and inline cases (e.g., “Chapter 1 2026”, “9 780113 227426”).
  - Resolves folios once at document scope and carries masks/pages into per‑page extraction; table cells are never removed; folio slots are excluded from column histograms and complexity scoring.
  - Markdown: drops lines like “Page 42”, “Page123”, “- 42 -”, while keeping inline numbers; adds `to_markdown_from_items_with_rects_and_page_count` to use the authoritative page count for coverage.
  - Forms: maps widget annotations to owning pages via the page tree to support page‑filtered extraction, including widgets that omit `/P`.
  - Adds regression tests for the evidence floor (including sparse/trailing blanks), spread/adjacent alternation, preserved inline digits, page‑filter selection, per‑page table safety, stabilized decisions, and widget page resolution.

<sup>Written for commit ef96c68e90b7650f29fc7fca7eaeb712709a2fe3. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/firecrawl/pdf-inspector/pull/201?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

File: src/extractor/layout.rs [hunk 1]
```diff
@@ -960,22 +960,141 @@ fn spans_multiple_columns(item: &TextItem, columns: &[ColumnRegion]) -> bool {
...
+    recurring
+}
+
+/// Decide which digit-only page-edge items can be removed before layout.
+///
+/// PDF producers commonly emit one text-showing operation per word. A numeric
+/// item attached to neighboring content on the same baseline is therefore kept,
+/// unless document-level recurrence identifies it as a running folio.
+fn page_number_removal_mask(items: &[TextItem]) -> Vec<bool> {
+    let recurring = recurring_page_number_mask(items);
```
[the remark this answers is not in the document]
Comment (abimaelmartell) [reply src/extractor/layout.rs]: Addressed in 489d0b9. Recurring folios are now identified from the complete document and removed before band or chart partitioning; standalone context filtering remains local to each resulting layout region.

File: src/extractor/layout.rs [hunk 2]
```diff
@@ -960,22 +960,141 @@ fn spans_multiple_columns(item: &TextItem, columns: &[ColumnRegion]) -> bool {
...
+            let has_context = (bucket - 1..=bucket + 1).any(|nearby_bucket| {
+                baseline_buckets
+                    .get(&(item.page, nearby_bucket))
+                    .is_some_and(|indices| {
+                        indices.iter().any(|&other_index| {
+                            if other_index == index {
+                                return false;
+                            }
+                            let other = &items[other_index];
+                            if other.text.trim().is_empty()
```
[the remark this answers is not in the document]
Comment (abimaelmartell) [reply src/extractor/layout.rs]: Addressed in 489d0b9. Same-baseline numeric candidates now form connected runs, and a run only establishes context when it also contains nonnumeric content. Numeric-only clusters no longer preserve one another; contextual chains such as Chapter 1 2026 remain intact.

**Comment by abimaelmartell** [comment]
Addressed all three review findings in 489d0b9:

- replaced the quadratic recurrence scan with spatial indexing and exact-geometry memoization
- carry document-level recurring folio decisions into split layout paths
- prevent numeric-only page-edge clusters from preserving one another while retaining real text context

Formatting, strict Clippy, the full Rust test suite, and the release build all pass. @cubic-dev-ai please re-review.

File: src/extractor/layout.rs [hunk 3]
```diff
@@ -960,22 +960,231 @@ fn spans_multiple_columns(item: &TextItem, columns: &[ColumnRegion]) -> bool {
...
+/// unless document-level recurrence identifies it as a running folio.
+fn page_number_removal_mask(items: &[TextItem]) -> Vec<bool> {
+    let candidate_values: Vec<Option<u32>> = items.iter().map(page_number_value).collect();
+    let recurring = recurring_page_number_mask(items, &candidate_values);
+    let contextual = page_number_context_mask(items, &candidate_values);
+
+    candidate_values
+        .iter()
+        .enumerate()
+        .map(|(index, value)| value.is_some() && (recurring[index] || !contextual[index]))
```
[the remark this answers is not in the document]
Comment (abimaelmartell) [reply src/extractor/layout.rs]: Valid—the short-document case cannot rely on recurrence. The updated code recognizes explicit page-number expressions before layout partitioning, so the numeric item in Page 42 and Page 3 of 10 is removed even on a single page. It keeps the total in Page N of M and does not delete a split Page 42 explains… run during layout; the established final Markdown cleanup still handles completed, isolated Page N running-header lines. Tests cover the adjacent short-document form, a running-header suffix, and partitioned pages. For an unlabeled contextual number, the P2 and P3 cases are geometrically indistinguishable, so the loss-averse behavior is to preserve it rather than guess.

File: src/extractor/layout.rs [hunk 4]
```diff
@@ -960,22 +960,231 @@ fn spans_multiple_columns(item: &TextItem, columns: &[ColumnRegion]) -> bool {
...
+                }
+            }
+        }
+
+        let mut page_values: Vec<(u32, u32)> = values_by_page
+            .into_iter()
+            .map(|(page, (_, value))| (page, value))
+            .collect();
+        page_values.sort_unstable_by_key(|&(page, _)| page);
+        let is_recurring = page_values.len() >= 3
```
[the remark this answers is not in the document]
Comment (abimaelmartell) [reply src/extractor/layout.rs]: Valid. I removed the recurrence detector entirely because an increasing numeric content column and an increasing folio have the same geometric signal. Contextual page-edge numbers are now preserved regardless of their cross-page sequence; only standalone candidates or explicit page-number syntax are removed. The new regression test uses 13, 14, and 15 beside Person across three pages. Preserved edge numbers are excluded only from column inference, then retained for line assembly, so preserving them does not perturb the existing layout decision.

**Comment by abimaelmartell** [comment]
@cubic-dev-ai The latest push addresses both findings: generic numeric recurrence is no longer a deletion signal, increasing contextual sequences are covered by regression tests, and explicit Page N forms are handled for short documents without broadening pre-layout deletion. Please review again.

**Comment by abimaelmartell** [comment]
Handled the remaining running-folio edge cases: decorative markers no longer establish lexical context, repeated advancing margin slots are filtered conservatively, and facing-page folio pairs are recognized geometrically. @cubic-dev-ai

**Review by abimaelmartell — COMMENTED** [review]
Previous review responses.

File: src/extractor/layout.rs [hunk 5]
```diff
@@ -960,22 +960,345 @@ fn spans_multiple_columns(item: &TextItem, columns: &[ColumnRegion]) -> bool {
...
+                .collect();
+            values.sort_by_key(|(page, _, _)| *page);
+
+            let unique_values: HashSet<u32> = values.iter().map(|(_, value, _)| *value).collect();
+            let mostly_unique = unique_values.len() * 5 >= values.len() * 4;
+            let increasing_pairs = values
+                .windows(2)
+                .filter(|pair| pair[1].1 > pair[0].1)
+                .count();
+            let mostly_increasing = increasing_pairs * 5 >= (values.len() - 1) * 4;
```
[the remark this answers is not in the document]
Comment (abimaelmartell) [reply src/extractor/layout.rs]: Addressed in 4252959. Repeated contextual candidates now have to track page deltas at a folio-like 1x or 2x rate and remain plausible for the document length. Added coverage preserving the 2020-2023 Annual report sequence.

File: src/extractor/layout.rs [hunk 6]
```diff
@@ -960,22 +960,345 @@ fn spans_multiple_columns(item: &TextItem, columns: &[ColumnRegion]) -> bool {
...
+                continue;
+            }
+            let Some(value) = candidate_values[index] else {
+                continue;
+            };
+            let paired = known_folios.iter().any(|&other| {
+                let other_value = candidate_values[other].unwrap();
+                let same_baseline =
+                    (items[index].y - items[other].y).abs() < PAGE_NUMBER_Y_TOLERANCE;
+                let opposite_sides = (items[index].x - items[other].x).abs()
```
[the remark this answers is not in the document]
Comment (abimaelmartell) [reply src/extractor/layout.rs]: Addressed in 4252959. A spread pair now requires a genuinely wide content span and candidates in opposing outer page quarters, in addition to the baseline and consecutive-value checks. Added coverage retaining contextual 12 Total beside an unrelated 13.

File: src/extractor/layout.rs [hunk 7]
```diff
@@ -960,22 +960,345 @@ fn spans_multiple_columns(item: &TextItem, columns: &[ColumnRegion]) -> bool {
...
+                        .map(|&index| items[index].text.trim())
+                        .filter(|text| !text.is_empty())
+                        .collect::<Vec<_>>()
+                        .join(" ");
+                    let candidates: Vec<(usize, u32)> = group
+                        .iter()
+                        .filter_map(|&index| candidate_values[index].map(|value| (index, value)))
+                        .collect();
+                    let in_deep_margin = candidates.iter().all(|(index, _)| {
+                        items[*index].y < REPEATED_FOLIO_MARGIN_Y
```
[the remark this answers is not in the document]
Comment (abimaelmartell) [reply src/extractor/layout.rs]: Addressed in 4252959. Candidate detection and repeated-folio classification now share the same page-edge bounds. Added coverage for a contextual footer at y=80.

**Comment by abimaelmartell** [comment]
Addressed all three findings in 4252959:

- repeated contextual candidates must track page progression and remain plausible for the document
- spread pairs require opposing outer edges of a genuinely wide content span
- recurrence uses the same page-edge band as candidate detection

Each case has regression coverage. @cubic-dev-ai please re-review.

File: src/extractor/layout.rs [hunk 8]
```diff
@@ -960,22 +960,377 @@ fn spans_multiple_columns(item: &TextItem, columns: &[ColumnRegion]) -> bool {
...
+/// the document, however, the surrounding text and Y position repeat while the
+/// numeric slot advances. Require that full signal before treating the slot as
+/// a folio so constant metadata and substantive rows near the page edge remain
+/// untouched.
+fn mark_repeated_folio_candidates(
+    occurrences_by_signature: HashMap<String, Vec<ContextualCandidateOccurrence>>,
+    document_page_count: usize,
+    explicit_folio: &mut [bool],
+) {
+    let min_pages = 4usize.max((document_page_count * 30).div_ceil(100));
```
[the remark this answers is not in the document]
Comment (abimaelmartell) [reply src/extractor/layout.rs:1013]: The max is intentional: both the four-occurrence floor and proportional document coverage are required. With min, one occurrence would satisfy a one-page document and could delete an ordinary contextual margin number. b2e2028 documents that invariant and adds a four-page/two-occurrence regression that the suggested change would fail.

**Comment by abimaelmartell** [comment]
Reviewed the remaining threshold suggestion in b2e2028. The existing `max` is intentional: both the four-occurrence evidence floor and proportional document coverage are required; using `min` would make one occurrence sufficient in a one-page document. The invariant is now documented and covered by a sparse-recurrence regression. @cubic-dev-ai please re-review.

**Comment by abimaelmartell** [comment]
Addressed all three findings in c60a0f3:

- the pre-partition test now asserts the filter boundary directly
- recurrence coverage uses the document page range rather than text-bearing page count
- page-number decisions are made once before partitioning and carried through every split path

Added regression coverage for sparse pages and context split across partitions. @cubic-dev-ai please re-review.

**Comment by abimaelmartell** [comment]
Addressed both findings in adc5415:

- Markdown now carries the authoritative PDF page count into its single document-level folio decision, including trailing blank pages
- numeric-only partitions provide no column evidence and safely use the single-column fallback

Added focused regressions for a 20-page sparse document and a numeric-only two-column-shaped partition. All Rust checks pass. @cubic-dev-ai please re-review.

**Comment by abimaelmartell** [comment]
Addressed the three latest findings in 2be669b:

- per-page Markdown extraction now applies document-wide folio decisions before partitioning
- lone long numeric tokens no longer preserve adjacent short folios, while structured and dense numeric rows remain intact
- callers can provide an authoritative page count through a backward-compatible public overload

Added focused regressions for each behavior. All Rust checks pass. @cubic-dev-ai please re-review.

**Comment by abimaelmartell** [comment]
Addressed the valid layout-metadata and performance findings in 6bd28fb. The table/band suggestion was tested and bounded deliberately: structural table detection keeps original numeric evidence to avoid deleting legitimate cells or creating phantom tables, while folios are excluded from column inference and final non-table layout. The branch is also current with main, and all Rust checks pass. @cubic-dev-ai please re-review.

**Comment by abimaelmartell** [comment]
Addressed the valid per-page table-boundary finding in bae8d34 by carrying document-wide folio decisions as a mask over original table inputs, with focused regression coverage. The numeric-context suggestion was reviewed against the preservation contract and existing positive/negative regressions; its broader lexical-only rule would delete legitimate numeric-only lists and identifiers, so the narrower tested boundary remains. The branch is current with main and all Rust checks pass. @cubic-dev-ai please re-review.

**Comment by abimaelmartell** [comment]
Addressed the page-filter consistency finding in 4c604b7. Folio decisions now use complete-document context before selecting requested pages, with focused regressions covering layout metadata and page scoping. The branch is current with main and all Rust checks pass. @cubic-dev-ai please re-review.

**Comment by abimaelmartell** [comment]
Addressed the adjacent-page safety findings in c6f291a: edge inference now requires a genuinely wide font-relative content span, and a contextual candidate needs two independent pre-existing anchors on distinct pages in the same advancing sequence. The complete Page N of M counter remains intentionally removable so it cannot leave an orphan total or malformed residue; focused coverage preserves any real trailing header suffix. All Rust checks pass. @cubic-dev-ai please re-review.

**Comment by abimaelmartell** [comment]
@cubic-dev-ai please re-review.

**Comment by abimaelmartell** [comment]
@cubic-dev-ai please re-review.
